lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1491,10 +1491,15 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
     /// use or capture of "self." for qualification of member references.
     static bool isClosureRequiringSelfQualification(
                   const AbstractClosureExpr *CE) {
+      auto closureTy = CE->getType();
+      // If we don't have a type for this closure yet (or the closure has error
+      // type), don't offer diagnostics for required 'self'. We will have
+      // another chance when (if) the closure successfully typechecks.
+      if (!closureTy || closureTy->hasError())
```

Comment:
Do we need the `!closureTy`? I thought this code wouldn't be run at all if something failed.

Reply:
Before you merged #33116, the 'explicit `self`' check here was getting triggered before we had resolved the types of all containing closures. So AFAICT there's nothing actually preventing this check from happening before we have a type, other than ensuring `performSyntacticExprDiagnostics` doesn't get called before all containing closures have a type.

If you're satisfied that this is otherwise guaranteed somewhere up the stack, then yeah, I believe this check is redundant. Could also be changed to an assertion.

Reply:
@DougGregor Sorry for the ping, just want to close the loop on this PR. If you have the time to provide your preference among the following options I'd appreciate it:

- Keep the runtime check as it is in this PR since we don't have 'up the stack' checking that the necessary types are resolved before calling into these diagnostics.
- Change this check to an assertion.
- Keep the logic in `isClosureRequiringSelfQualification` as it is on `master` and add the additional regression test.
- Close out this PR without merging since #33116 sufficiently addresses the underlying issue.

I'm happy with all of the above, just don't want to let this PR sit around indefinitely and clog up the list of PRs!